### PR TITLE
[vSphere][datastorecluster] Add new Datastream Datastore Cluster

### PR DESCRIPTION
**Pull request by kush-elastic**
## Description
Here are the following metrics to be added as part of new datastream datastore cluster in the vSphere Integration.

Metrics Type | Metrics | API | Field Mappings
-- | -- | -- | --
Datastore cluster capacity | StoragePod.Summary.Capacity | Summary | capacity.bytes
  | StoragePod.Summary.FreeSpace | Summary | free_space.bytes
  Datastore cluster | childEntity(type= datastore) | Extra API | datastore.names
  | childEntity(type= datastore) | Extra API | datastore.count

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## TSDB

```
Values being used:
        elasticsearch_host = https://localhost:9200
        elasticsearch_ca_path = /root/.elastic-package/profiles/default/certs/elasticsearch/ca-cert.pem
        elasticsearch_user = elastic
        elasticsearch_pwd = changeme
        data_stream = metrics-vsphere.datastorecluster-default
        docs_index = First index of the data stream
        settings_mappings_index = Last index of the data stream
        max_docs = All documents
        get_overlapping_files = True
        directory_overlapping_files = overwritten-docs-metrics-vsphere.datastorecluster-default
        display_docs = 10
        copy_docs_per_dimension = 2

You're testing with version 8.15.2-SNAPSHOT.

Testing data stream metrics-vsphere.datastorecluster-default.
Index being used for the documents is .ds-metrics-vsphere.datastorecluster-default-2024.09.22-000001.
Index being used for the settings and mappings is .ds-metrics-vsphere.datastorecluster-default-2024.09.22-000001.

The time series fields for the TSDB index are: 
        - dimension (10 fields):
                - agent.id
                - cloud.account.id
                - cloud.availability_zone
                - cloud.instance.id
                - cloud.provider
                - cloud.region
                - container.id
                - host.name
                - service.address
                - vsphere.datastorecluster.name
        - gauge (3 fields):
                - vsphere.datastorecluster.capacity.bytes
                - vsphere.datastorecluster.datastore.count
                - vsphere.datastorecluster.free_space.bytes
        - routing_path (10 fields):
                - agent.id
                - cloud.account.id
                - cloud.availability_zone
                - cloud.instance.id
                - cloud.provider
                - cloud.region
                - container.id
                - host.name
                - service.address
                - vsphere.datastorecluster.name

Index tsdb-index-enabled successfully created.

Copying documents from .ds-metrics-vsphere.datastorecluster-default-2024.09.22-000001 to tsdb-index-enabled...
All 6 documents taken from index .ds-metrics-vsphere.datastorecluster-default-2024.09.22-000001 were successfully placed to index tsdb-index-enabled.
```

## Related issues

- Closes https://github.com/elastic/obs-integration-team/issues/45

### CODE REVIEW

File: packages/vsphere/data_stream/datastorecluster/fields/fields.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      type: group
+      fields:
+        - name: names
+          type: keyword
+          description: >
+            List of all the datastore names.
+        - name: count
+          type: long
+          description: >
+            Number of datastores in the Datastore Cluster.
```
Comment (lucian-ioan) [inline packages/vsphere/data_stream/datastorecluster/fields/fields.yml:32]: Add metric_type to all these fields.

File: packages/vsphere/data_stream/datastorecluster/manifest.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+          - vsphere-datastorecluster
+      - name: processors
+        type: yaml
+        title: Processors
+        multi: false
+        required: false
+        show_user: false
+        description: Processors are used to reduce the number of fields in the exported event or to enhance the event with metadata. This executes in the agent before the logs are parsed. See [Processors](https://www.elastic.co/guide/en/beats/filebeat/current/filtering-and-enhancing-data.html) for details.
+elasticsearch:
+  index_mode: "time_series"
```
Comment (lucian-ioan) [inline packages/vsphere/data_stream/datastorecluster/manifest.yml:24]: Please provide a screenshot with results from [TSDB-migration-test-kit](https://github.com/elastic/TSDB-migration-test-kit) before enabling TSDB (after adding metric_type to the fields).

**Review by lucian-ioan — CHANGES_REQUESTED** [review]
Add a changelog entry with version "1.14.0-next" as per this [issue](https://github.com/elastic/obs-integration-team/issues/22).

File: packages/vsphere/data_stream/datastorecluster/fields/fields.yml
Comment (niraj-elastic) [inline packages/vsphere/data_stream/datastorecluster/fields/fields.yml:1]: Lets add triggered alarms as well.

File: packages/vsphere/data_stream/datastorecluster/fields/agent.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  description: 'A host is defined as a general computing instance.  ECS host.* fields should be populated with details about the host on which the event happened, or from which the measurement was taken. Host types include hardware, virtual machines, Docker containers, and Kubernetes nodes.'
+  type: group
+  fields:
+    - name: name
+      level: core
+      type: keyword
+      dimension: true
+      ignore_above: 1024
+      description: 'Name of the host.  It can contain what `hostname` returns on Unix systems, the fully qualified domain name, or a name specified by the user. The sender decides which value to use.'
+    - name: containerized
```
Comment (niraj-elastic) [inline packages/vsphere/data_stream/datastorecluster/fields/agent.yml:13]: I think ecs@mapping could take care of this fields, if these fields are included in ecs and are not dimension than we dont need to mention these.

**Review by niraj-elastic — CHANGES_REQUESTED** [review]

File: packages/vsphere/_dev/build/docs/README.md [hunk 4]
```diff
@@ -103,4 +103,26 @@ This is the `virtualmachine` data stream. This data stream collects virtual mach
 
 Please refer to the following [document](https://www.elastic.co/guide/en/ecs/current/ecs-field-reference.html) for detailed information on ECS fields.
 
-{{fields "virtualmachine"}}
\ No newline at end of file
+{{fields "virtualmachine"}}
+
+### Datastore Cluster
```
Comment (ishleenk17) [inline packages/vsphere/_dev/build/docs/README.md:108]: We should also add this to the datastreams section.

File: packages/vsphere/_dev/build/docs/README.md [hunk 5]
```diff
@@ -103,4 +103,26 @@ This is the `virtualmachine` data stream. This data stream collects virtual mach
...
+
+Datastore clusters in vSphere group multiple datastores for optimized management and automated load balancing, ensuring efficient storage utilization and simplified administration.
+
+**ECS Field Reference**
+
+Please refer to the following [document](https://www.elastic.co/guide/en/ecs/current/ecs-field-reference.html) for detailed information on ECS fields.
+
+{{fields "datastorecluster"}}
+
+## Logs
```
Comment (ishleenk17) [inline packages/vsphere/_dev/build/docs/README.md:118]: Why is this being added in this ? I thought we already had this section

File: packages/vsphere/data_stream/datastorecluster/fields/fields.yml (hunk 1 above)
Comment (kush-elastic) [reply packages/vsphere/data_stream/datastorecluster/fields/fields.yml]: Updated

File: packages/vsphere/data_stream/datastorecluster/fields/fields.yml
Comment (kush-elastic) [reply packages/vsphere/data_stream/datastorecluster/fields/fields.yml:1]: Added

File: packages/vsphere/data_stream/datastorecluster/fields/agent.yml (hunk 3 above)
Comment (kush-elastic) [reply packages/vsphere/data_stream/datastorecluster/fields/agent.yml]: removed unnecessary

File: packages/vsphere/_dev/build/docs/README.md (hunk 4 above)
Comment (kush-elastic) [reply packages/vsphere/_dev/build/docs/README.md]: Right

File: packages/vsphere/_dev/build/docs/README.md [hunk 6]
```diff
@@ -116,3 +128,7 @@ This is the `virtualmachine` data stream. This data stream collects virtual mach
 Please refer to the following [document](https://www.elastic.co/guide/en/ecs/current/ecs-field-reference.html) for detailed information on ECS fields.
 
 {{fields "virtualmachine"}}
+<<<<<<< HEAD
```
Comment (niraj-elastic) [inline packages/vsphere/_dev/build/docs/README.md:131]: please fix this.

File: packages/vsphere/_dev/build/docs/README.md [hunk 7]
```diff
@@ -116,3 +128,7 @@ This is the `virtualmachine` data stream. This data stream collects virtual mach
...
+
+=======
+>>>>>>> bde6b5118be928daf948a9cdfa1eadb0a857a852
```
Comment (ishleenk17) [inline packages/vsphere/_dev/build/docs/README.md:134]: Left out from a merge ?

File: packages/vsphere/_dev/deploy/docker/Dockerfile [hunk 8]
```diff
@@ -6,4 +6,4 @@ RUN apk add --no-cache curl git
 RUN go install github.com/vmware/govmomi/vcsim@v0.30.4
 
 HEALTHCHECK --interval=1s --retries=60 --timeout=10s CMD curl http://localhost:8989/
-CMD vcsim -l :8989
+CMD vcsim -pod 1 -l :8989
```
Comment (ishleenk17) [inline packages/vsphere/_dev/deploy/docker/Dockerfile:9]: Why the change here ?

File: packages/vsphere/data_stream/datastorecluster/manifest.yml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+type: metrics
+streams:
+  - input: vsphere/metrics
+    title: vSphere Datastore Cluster metrics
+    description: vSphere Datastore Cluster metrics
+    vars:
+      - name: period
+        type: text
+        title: Period
+        description: The period for fetching metrics. Supported units for this parameter are s (seconds), h (hours) and m (minutes).
```
Comment (ishleenk17) [inline packages/vsphere/data_stream/datastorecluster/manifest.yml:11]: Description differes from other datastreams.

**Review by niraj-elastic — CHANGES_REQUESTED** [review]

File: packages/vsphere/_dev/build/docs/README.md (hunk 7 above)
Comment (kush-elastic) [reply packages/vsphere/_dev/build/docs/README.md]: Fixed.

File: packages/vsphere/_dev/deploy/docker/Dockerfile (hunk 8 above)
Comment (kush-elastic) [reply packages/vsphere/_dev/deploy/docker/Dockerfile:9]: vcsim's default instance does not have any StoragePod.

File: packages/vsphere/_dev/build/docs/README.md (hunk 6 above)
Comment (kush-elastic) [reply packages/vsphere/_dev/build/docs/README.md]: Fixed.

**Comment by willemdh** [comment]
Is there support for Vsphere 8?

File: packages/vsphere/data_stream/datastorecluster/manifest.yml (hunk 9 above)
Comment (kush-elastic) [reply packages/vsphere/data_stream/datastorecluster/manifest.yml]: fixed

**Comment by kush-elastic** [comment]
> Is there support for Vsphere 8?

We have tested current integration with version 7.0.3, i cannot say confidently but it might just work fine with 8.0.0 as well.

File: packages/vsphere/data_stream/datastorecluster/manifest.yml (hunk 2 above)
Comment (kush-elastic) [reply packages/vsphere/data_stream/datastorecluster/manifest.yml:29]:
```
Values being used:
        elasticsearch_host = https://localhost:9200
        elasticsearch_ca_path = /root/.elastic-package/profiles/default/certs/elasticsearch/ca-cert.pem
        elasticsearch_user = elastic
        elasticsearch_pwd = changeme
        data_stream = metrics-vsphere.datastorecluster-default
        docs_index = First index of the data stream
        settings_mappings_index = Last index of the data stream
        max_docs = All documents
        get_overlapping_files = True
        directory_overlapping_files = overwritten-docs-metrics-vsphere.datastorecluster-default
        display_docs = 10
        copy_docs_per_dimension = 2

You're testing with version 8.15.2-SNAPSHOT.

Testing data stream metrics-vsphere.datastorecluster-default.
Index being used for the documents is .ds-metrics-vsphere.datastorecluster-default-2024.09.22-000001.
Index being used for the settings and mappings is .ds-metrics-vsphere.datastorecluster-default-2024.09.22-000001.

The time series fields for the TSDB index are: 
        - dimension (10 fields):
                - agent.id
                - cloud.account.id
                - cloud.availability_zone
                - cloud.instance.id
                - cloud.provider
                - cloud.region
                - container.id
                - host.name
                - service.address
                - vsphere.datastorecluster.name
        - gauge (3 fields):
                - vsphere.datastorecluster.capacity.bytes
                - vsphere.datastorecluster.datastore.count
                - vsphere.datastorecluster.free_space.bytes
        - routing_path (10 fields):
                - agent.id
                - cloud.account.id
                - cloud.availability_zone
                - cloud.instance.id
                - cloud.provider
                - cloud.region
                - container.id
                - host.name
                - service.address
                - vsphere.datastorecluster.name

Index tsdb-index-enabled successfully created.

Copying documents from .ds-metrics-vsphere.datastorecluster-default-2024.09.22-000001 to tsdb-index-enabled...
All 6 documents taken from index .ds-metrics-vsphere.datastorecluster-default-2024.09.22-000001 were successfully placed to index tsdb-index-enabled.
```

**Review by niraj-elastic — APPROVED** [review]
LGTM

**Review by lucian-ioan — APPROVED** [review]
LGTM

File: packages/vsphere/_dev/deploy/docker/docker-compose.yml [hunk 10]
```diff
@@ -1,7 +1,6 @@
 version: '2.3'
 services:
   vsphere-metrics:
-    image: docker.elastic.co/integrations-ci/beats-vsphere:${VSPHERE_GOLANG_VERSION:-1.17}-1
```
Comment (ishleenk17) [inline packages/vsphere/_dev/deploy/docker/docker-compose.yml:4]: IS this image not used ?

File: packages/vsphere/data_stream/datastorecluster/fields/agent.yml [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      type: boolean
+      description: >
+        If the host is a container.
+
+    - name: os.build
```
Comment (ishleenk17) [inline packages/vsphere/data_stream/datastorecluster/fields/agent.yml:18]: Are os.build and os.codename ecs fields ?

File: packages/vsphere/data_stream/datastorecluster/manifest.yml [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+        show_user: false
+        default:
+          - vsphere-datastorecluster
+      - name: processors
+        type: yaml
+        title: Processors
+        multi: false
+        required: false
+        show_user: false
+        description: Processors are used to reduce the number of fields in the exported event or to enhance the event with metadata. This executes in the agent before the logs are parsed. See [Processors](https://www.elastic.co/guide/en/beats/filebeat/current/filtering-and-enhancing-data.html) for details.
```
Comment (ishleenk17) [inline packages/vsphere/data_stream/datastorecluster/manifest.yml:27]: Link to the Fleet Processors pls

File: packages/vsphere/_dev/deploy/docker/docker-compose.yml (hunk 10 above)
Comment (kush-elastic) [reply packages/vsphere/_dev/deploy/docker/docker-compose.yml:4]: it was using image by default. if we want to use dockerfile we need to remove it.

File: packages/vsphere/data_stream/datastorecluster/fields/agent.yml (hunk 11 above)
Comment (kush-elastic) [reply packages/vsphere/data_stream/datastorecluster/fields/agent.yml:18]: i think so, this i followed from datastore datastream, i think @lucian-ioan might be able to explain this.

File: packages/vsphere/data_stream/datastorecluster/manifest.yml (hunk 12 above)
Comment (kush-elastic) [reply packages/vsphere/data_stream/datastorecluster/manifest.yml:27]: sure

Comment (kush-elastic) [reply packages/vsphere/data_stream/datastorecluster/manifest.yml]: Fixed

File: packages/vsphere/data_stream/datastorecluster/fields/agent.yml (hunk 11 above)
Comment (ishleenk17) [reply packages/vsphere/data_stream/datastorecluster/fields/agent.yml:18]: If its ecs and not dimension field, we should remove it.

**Review by ishleenk17 — APPROVED** [review]
Approving, please address 1 pending comment.

File: packages/vsphere/data_stream/datastorecluster/fields/agent.yml (hunk 11 above)
Comment (kush-elastic) [reply packages/vsphere/data_stream/datastorecluster/fields/agent.yml]: Removed

**Pull request merged**